larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -590,19 +593,138 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
+
+        return this->GetClosestPfo(pPfoToEnlarge, reducedPfoVector);
+    }
+    else
+    {
+        for (const ParticleFlowObject *const pPfoToShift : pfoVector) reducedPfoVector.push_back(pPfoToShift);
+        std::sort(reducedPfoVector.begin(), reducedPfoVector.end(), LArPfoHelper::SortByNHits);
+
+        return pPfoToEnlarge;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const ParticleFlowObject *StitchingCosmicRayMergingTool::GetClosestPfo(const ParticleFlowObject *const pPfoToEnlarge, const PfoVector &pfoVector) const
+{
+    float minDistance(std::numeric_limits<float>::max());
+    for (const ParticleFlowObject *const pPfoToShift : pfoVector)
+    {
+        if (pPfoToShift == pPfoToEnlarge)
+            return pPfoToEnlarge;
+
+        const float thisDistance(LArPfoHelper::GetTwoDSeparation(pPfoToShift, pPfoToEnlarge));
+        if (thisDistance < minDistance)
+            minDistance = thisDistance;
+    }
+
+    //TODO - better way to avoid this loop? maybe create a map pfo,distance?
+    for (const ParticleFlowObject *const pPfoToShift : pfoVector)
+    {
+        if ((LArPfoHelper::GetTwoDSeparation(pPfoToShift, pPfoToEnlarge) - minDistance) < std::numeric_limits<float>::epsilon())
+            return pPfoToShift;
+    }
+
+    //ATTN: should not arrive here, but if we did, return the first element in pfoVector
+    PfoVector::const_iterator iter = pfoVector.begin();
+    return *iter;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+PfoVector StitchingCosmicRayMergingTool::SelectLongestStitch(PfoVector &pfoVector, const PfoToLArTPCMap &pfoToLArTPCMap) const
+{
+    PfoVector selectedPfos;
+    int totalHitsStitched(0);
+
+    for (const ParticleFlowObject *const pPfoToShift1 : pfoVector)
+    {
+        for (const ParticleFlowObject *const pPfoToShift2 : pfoVector)
+        {
+            if (pPfoToShift2 == pPfoToShift1)
+                continue;
+
+            const unsigned int twoDHitsPfo1(this->NumberOfTwoDHits(pPfoToShift1)), twoDHitsPfo2(this->NumberOfTwoDHits(pPfoToShift2));
+
+            PfoToLArTPCMap::const_iterator iter1(pfoToLArTPCMap.find(pPfoToShift1));
+            PfoToLArTPCMap::const_iterator iter2(pfoToLArTPCMap.find(pPfoToShift2));
+
+            if ((iter1 != pfoToLArTPCMap.end()) && (iter2 != pfoToLArTPCMap.end()))
+            {
+                const LArTPC *const pLArTPC1(iter1->second);
+                const LArTPC *const pLArTPC2(iter2->second);
+
+                if (((twoDHitsPfo1 + twoDHitsPfo2) > totalHitsStitched) && LArStitchingHelper::CanTPCsBeStitched(*pLArTPC1, *pLArTPC2))
+                {
+                    totalHitsStitched = twoDHitsPfo1 + twoDHitsPfo2;
+                }
+            }
+        }
+    }
+
```

Comment:
Almost as above, typical approach is to declare a new PfoVector at start and initialise it to your default return (so first two pfos from input pfoVector). Then have a single loop where you keep track of `totalHitsStitched`, and assign to vector whenever you find a stitch involving more hits. After the single loop, can just return the new vector.

Note, we don't often return a vector by value in Pandora, as it's a copy at the end, but you can do this for cleanliness and clarity here if you like, as these will not be big vectors. (Alternative would be to pass by non-const reference, of course).

Reply:
implemented, thanks, it looks nicer now :) I have also avoided the return vector by value and instead passed as argument and edit it inside